### PR TITLE
Add support for CLI tool target

### PR DESCRIPTION
**Pull request by olejnjak**
Resolves https://github.com/tuist/tuist/issues/1224

### Short description 📝

Adds support for generating CLI tool targets. 

### Solution 📦

Just added `commandLineTool` as `Product` type and altered related properties and tests.

### Implementation 👩‍💻👨‍💻

- [x] add `commandLineTool` as `Product` type
- [x] fix missing cases errors generated by compiler
- [x] update `runnable` and `forPlatform(_:)` bodies
- [x] update related tests

### CODE REVIEW

File: Sources/TuistSigning/ProvisioningProfile/ProvisioningProfile.swift [hunk 1]
```diff
@@ -3,7 +3,7 @@ import TSCBasic
 import TuistSupport
 
 /// Model of a provisioning profile
-struct ProvisioningProfile: Equatable {
+public struct ProvisioningProfile: Equatable {
```
Comment (fortmarek) [inline Sources/TuistSigning/ProvisioningProfile/ProvisioningProfile.swift:6]: This probably should be kept `internal`?

**Review by fortmarek — COMMENTED** [review]
Thanks for doing this! I'd just maybe add a fixture, so we can ensure this works properly in an acceptance test?

File: Sources/TuistSigning/ProvisioningProfile/ProvisioningProfile.swift (hunk 1 above)
Comment (olejnjak) [reply Sources/TuistSigning/ProvisioningProfile/ProvisioningProfile.swift]: ✅

**Review by fortmarek — APPROVED** [review]

**Review by kwridan — COMMENTED** [review]
Thanks for adding this @olejnjak 👍 

The changes made here are the right ones to create command line tool targets in isolation, but if I'm not mistaken won't support linking dependencies. For that some of the `Graph` and `GraphLinter` logic would need to be updated to support new linking rules etc... 

That said, to keep PRs small, this change is fine as is the linter will warn us that dependencies are not supported if they are defined :) 

Some minor suggestions:

- We can update the documentation to mention the new supported product type
- We can also add a small fixture in `fixtures/` to help us test / show case this feature, especially if we do later revisit it to add dependency support

**Review by pepibumur — COMMENTED** [review]
Thank you very much for contributing to this @olejnjak. My feedback is similar to @kwridan's. - I'd extend [this linting logic](https://github.com/tuist/tuist/blob/master/Sources/TuistGenerator/Linter/GraphLinter.swift#L256) to account for command line tools. Those linters are very useful for users because they error if they are trying to define dependencies that are valid.
- I'd also add a new page to our "examples" section of the documentation: "Command line tools". Where we include an example of how a command line tool can be defined in Tuist. You can add the link [here](https://github.com/tuist/tuist/blob/master/website/markdown/docs/links.mdx), and add the page [here](https://github.com/tuist/tuist/tree/master/website/markdown/docs/examples).
- And I'd also add a simple acceptance test where there's a project with a CLI target, and I'd test that the target compiles. You can read more about testing [here](https://tuist.io/docs/contribution/testing-strategy/).

<img width="350" alt="image" src="https://user-images.githubusercontent.com/663605/97157812-e5920b00-1778-11eb-9cb8-d9adf698fafc.png"

**Comment by pepibumur** [comment]
@all-contributors add @olejnjak for idea

**Comment by olejnjak** [comment]
Well I think everything should be addressed - added test fixtures, some more unit tests, added docs.

Regarding dependencies, static libs work for me after archiving and static frameworks should work the same way, so adding these links to `validLinks` in `GraphLinter` should be sufficient, I haven't thought about any other linting validations that would need additional linting code as it did e.g. for app clips.

File: fixtures/command_line_tool_with_static_dependencies/Project.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import ProjectDescription
+
+let project = Project(name: "CommandLineTool",
+                      targets: [
+                        Target(name: "CommandLineTool",
+                               platform: .macOS,
+                               product: .commandLineTool,
+                               bundleId: "com.example.commandlinetool",
```
Comment (kwridan) [inline fixtures/command_line_tool_with_static_dependencies/Project.swift:8]: It's worth adding a few source files with compilable code - e.g. in the command line target importing and using code from the static library.

e.g. I added a `main.swift` and a `MyStaticLibClass.swift`, and that helped me discover we were missing linking static libraries 😅

**Review by kwridan — COMMENTED** [review]
Thanks for the updates @olejnjak

https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Models/Target.swift#L111 needs a small update to allow command line tools to link static targets during generation, otherwise we sadly get undefined symbol errors when compiling.

Screenshot 2020-11-03 at 07 09 07

**Review by kwridan — APPROVED** [review]
thanks @olejnjak

**Review by fortmarek — APPROVED** [review]
Looks good 👏

**Comment by fortmarek** [comment]
We can merge this when the master is fixed, is there some progress on that @pepibumur ?

File: features/generate-7.feature [hunk 3]
```diff
@@ -0,0 +1,21 @@
+Feature: Generate a new project using Tuist (suite 7)
+
+Scenario: The project is an macOS command line tool without any dependencies (command_line_tool_basic)
+    Given that tuist is available
+    And I have a working directory
+    Then I copy the fixture command_line_tool_basic into the working directory
+    Then tuist generates the project
+    Then I should be able to build for macOS the scheme CommandLineTool
```
Comment (pepibumur) [inline features/generate-7.feature:8]: @olejnjak one final suggestion I'd make here is: what about adding the following extra step:
```
Then I should be able to run the tool X
```
With that, we'd test that we've got a valid binary with all its dependencies linked.

File: website/markdown/docs/examples/command-line-tools.mdx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+---
+name: Command Line Tools
+excerpt: 'This page documents how to declare and use Command Line Tool targets.'
+---
+
+import Message from '../components/message'
+
+# Command Line Tools
+
+On macOS we can build various command line tools (like tuist) this way you can define your project.
```
Comment (pepibumur) [inline website/markdown/docs/examples/command-line-tools.mdx:10]:
```suggestion
Tuist supports defining CLI tools that might contain static dependencies. Below there's an example of how you'd define a project with a CLI target:
```

File: website/markdown/docs/examples/command-line-tools.mdx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    name: "CLITool",
+    targets: [
+        Target(
+            name: "CLITool",
+            platform: .macOS,
+            product: .commandLineTool,
+            bundleId: "io.tuist.App",
+            infoPlist: .default,
+            sources: ["Sources/**"],
+        ),
```
Comment (pepibumur) [inline website/markdown/docs/examples/command-line-tools.mdx:23]: I'd extend it to have a library as a dependency. That way users will see that it's possible to define dependencies for CLI too.

File: website/markdown/docs/examples/command-line-tools.mdx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    ]
+)
+```
+
+Keep in mind that command line tools support only statically linked dependencies, Tuist will do its best to keep his eye on that.
```
Comment (pepibumur) [inline website/markdown/docs/examples/command-line-tools.mdx:28]: We can remove this sentence.

File: website/markdown/docs/examples/command-line-tools.mdx [hunk 7]
```diff
@@ -0,0 +1,28 @@
+---
+name: Command Line Tools
+excerpt: 'This page documents how to declare and use Command Line Tool targets.'
```
Comment (pepibumur) [inline website/markdown/docs/examples/command-line-tools.mdx:3]:
```suggestion
excerpt: Learn how to use Tuist to define a CLI tools that might contain static dependencies.
```

**Review by pepibumur — CHANGES_REQUESTED** [review]
@olejnjak I left some minor comments. After addressing them and rebasing from master I thin we are good to merge. Thank you very much for contributing to this 🚀

File: website/markdown/docs/examples/command-line-tools.mdx (hunk 4 above)
Comment (olejnjak) [reply website/markdown/docs/examples/command-line-tools.mdx]: ✅

File: website/markdown/docs/examples/command-line-tools.mdx (hunk 6 above)
Comment (olejnjak) [reply website/markdown/docs/examples/command-line-tools.mdx]: ✅

File: website/markdown/docs/examples/command-line-tools.mdx (hunk 7 above)
Comment (olejnjak) [reply website/markdown/docs/examples/command-line-tools.mdx]: ✅

File: website/markdown/docs/examples/command-line-tools.mdx (hunk 5 above)
Comment (olejnjak) [reply website/markdown/docs/examples/command-line-tools.mdx:32]: ✅

File: features/generate-7.feature (hunk 3 above)
Comment (olejnjak) [reply features/generate-7.feature:8]: Okay, but that will take a while as I have literally no idea how to do that now...

**Review by pepibumur — APPROVED** [review]

**Comment by raven** [comment]
Thank you so much @olejnjak!

**Pull request merged**